Docs/background tasks (#15843)
mastra-ai/mastra: docs/src/content/en/docs/agents/background-tasks.mdx

---
title: "Background tasks | Agents"
description: Learn how to dispatch long-running tool calls in the background, keep the stream open until they complete, and orchestrate sub-agents asynchronously.
packages:
  - "@mastra/core"
  - "@mastra/memory"
---

# Background tasks

> 

Reviewer: Please add a ``**Added in:** `package@version` `` here so one knows how they can use it

Author: thanks added this

> Background tasks let an agent dispatch a long-running tool call without blocking the agentic loop. The tool returns an immediate acknowledgement, the LLM continues responding, and the task runs to completion in the background. When it finishes, its result is written to memory and if you use `streamUntilIdle()` the agent is re-invoked automatically so the result is processed in the same call.

Reviewer: Link the `streamUntilIdle()` to the reference page

## When to use background tasks

Use background tasks when a tool call may take long enough that the user shouldn't wait for it before seeing a response. Common cases:

- Sub-agent delegations that themselves run multi-step research or writing.
- Tool calls that hit slow external services, queues, or large data jobs.
- Workflows triggered from a tool call that may take minutes to complete.

> For tool calls that return quickly, foreground execution is simpler — keep using `agent.stream()` and `agent.generate()` directly.

Reviewer: ```suggestion
For tool calls that return quickly, foreground execution using `agent.stream()` and `agent.generate()` is simpler.
```

:::note

Background tasks require a configured [storage](/docs/memory/storage) backend on the Mastra instance. Tasks are persisted so they survive process restarts.

:::

> ## Enable on the Mastra instance

Reviewer: ```suggestion
## Quickstart
```

Background tasks are off by default. Enable them by setting `backgroundTasks.enabled` on the Mastra instance:

```typescript title="src/mastra/index.ts"
import { Mastra } from '@mastra/core'
import { LibSQLStore } from '@mastra/libsql'

export const mastra = new Mastra({
  agents: {
    /* ... */
  },
  storage: new LibSQLStore({ id: 'storage', url: 'file:mastra.db' }),
  backgroundTasks: {
    enabled: true,
    globalConcurrency: 10,
    perAgentConcurrency: 5,
    backpressure: 'queue',
    defaultTimeoutMs: 300_000,
  },
})

> ```

Reviewer: ```suggestion
```typescript title="src/mastra/index.ts" {6-12}
import { Mastra } from '@mastra/core'
import { LibSQLStore } from '@mastra/libsql'

export const mastra = new Mastra({
  storage: new LibSQLStore({ id: 'storage', url: 'file:mastra.db' }),
  backgroundTasks: {
    enabled: true,
    globalConcurrency: 10,
    perAgentConcurrency: 5,
    backpressure: 'queue',
    defaultTimeoutMs: 300_000,
  },
})
```
```

- Tool calls that hit slow external services, queues, or large data jobs.
- Workflows triggered from a tool call that may take minutes to complete.

For tool calls that return quickly, foreground execution is simpler — keep using `agent.stream()` and `agent.generate()` directly.

:::note

Background tasks require a configured [storage](/docs/memory/storage) backend on the Mastra instance. Tasks are persisted so they survive process restarts.

:::

## Enable on the Mastra instance

Background tasks are off by default. Enable them by setting `backgroundTasks.enabled` on the Mastra instance:

```typescript title="src/mastra/index.ts"
import { Mastra } from '@mastra/core'
import { LibSQLStore } from '@mastra/libsql'

export const mastra = new Mastra({
  agents: {
    /* ... */
  },
  storage: new LibSQLStore({ id: 'storage', url: 'file:mastra.db' }),
  backgroundTasks: {
    enabled: true,
    globalConcurrency: 10,
    perAgentConcurrency: 5,
    backpressure: 'queue',
    defaultTimeoutMs: 300_000,
  },
})
```

The full set of options is listed in the [backgroundTasks configuration reference](/reference/configuration#backgroundtasks).

> ## Opt a tool into running in the background

Reviewer: ```suggestion
## Run a tool in the background
```

> Enabling the manager doesn't run anything in the background by itself — every tool defaults to foreground execution. You opt a tool in at one of three layers, in priority order:

Reviewer: ```suggestion
Enabling the manager doesn't run anything in the background by itself as every tool defaults to foreground execution. You can run a tool in the background at one of three layers, in priority order:
```

```typescript title="src/mastra/agents/researcher.ts"
import { Agent } from '@mastra/core/agent'

export const researcher = new Agent({
  id: 'researcher',
  instructions: 'You research topics and answer questions.',
  model: 'openai/gpt-5.4',
  tools: { researchTool, summarizeTool },
  // highlight-start
  backgroundTasks: {
    tools: {
      researchTool: { enabled: true, timeoutMs: 600_000 },
      summarizeTool: false,
    },
  },
  // highlight-end
})
```

Set `tools: 'all'` to opt in every tool the agent has.

### LLM per-call override

When a tool is registered on an agent that has background tasks enabled, the model can include a `_background` field in the tool arguments to override the resolved configuration for that specific call:

```json
{
  "topic": "solana",
  "_background": { "enabled": true, "timeoutMs": 900_000 }
}
```

The override is stripped from the arguments before the tool runs. All fields in `_background` are optional, the model only includes what it wants to override.

> ### Resolution order

Reviewer: This paragraph reads like a repetition of the intro so it should be folded into that

Use this pattern when you want a subagent to behave consistently in the background regardless of which supervisor invokes it. Use the supervisor-side opt-in (above) when you want to tune background behavior centrally per supervisor.

## Lifecycle callbacks

Each layer can register completion callbacks. They all fire on completion and don't replace one another:

- Tool-level `backgroundTasks.onComplete` / `onFailed`: scoped to one tool.
- Agent-level `backgroundTasks.onTaskComplete` / `onTaskFailed`: scoped to all tasks dispatched by this agent.
- Manager-level `onTaskComplete` / `onTaskFailed`: scoped globally.

```typescript title="src/mastra/index.ts"
export const mastra = new Mastra({
  storage,
  backgroundTasks: {
    enabled: true,
    onTaskComplete: task => {
      logger.info('Background task complete', { taskId: task.id, toolName: task.toolName })
    },
    onTaskFailed: task => {
      logger.error('Background task failed', { taskId: task.id, error: task.error })
    },
  },
})
```

## Related

> - [Agent.streamUntilIdle() reference](/reference/streaming/agents/streamUntilIdle)

Reviewer: ```suggestion
- [`Agent.streamUntilIdle()` reference](/reference/streaming/agents/streamUntilIdle)
```
